elements/lisk-tree/src/sparse_merkle_tree/sparse_merkle_tree.ts

```diff
@@ -35,8 +43,108 @@ export class SparseMerkleTree {
+		}
+
+		const { leftHash, rightHash } = getBranchData(data);
+
+		return new Branch(leftHash, rightHash);
+	}
+
+	public async update(key: Buffer, value: Buffer) {
+		if (!value) {
+			throw new Error('Value cannot be empty');
+		}
+
+		if (key.length !== this.keyLength) {
+			throw new Error(`Key is not equal to defined key length of ${this.keyLength}`);
+		}
+
+		const newLeaf = new Leaf(key, value);
+		const ancestorNodes: TreeNode[] = [];
+		let bottomNode: TreeNode = new Empty();
+		let currentNode = await this.getNode(this.rootHash);
+		const binaryKey = binaryExpansion(key, this.keyLength);
+		let h = 0;
+
+		while (!isLeaf(currentNode.hash)) {
+			// Append currentNode to ancestorNodes
+			ancestorNodes.push(currentNode);
+			const d = binaryKey[h];
+			if (+d === 0) {
+				currentNode = await this.getNode(currentNode.hash);
+			} else if (+d === 1) {
+				currentNode = await this.getNode(currentNode.hash);
+			}
+			h += 1;
+			// The currentNode is an empty default node or a leaf node
+		}
+		// The currentNode is an empty node
+		if (currentNode.hash.equals(EMPTY_HASH)) {
+			bottomNode = newLeaf;
+		} else if (currentNode instanceof Leaf && key.equals(currentNode.key)) {
+			bottomNode = newLeaf;
+		} else {
+			// We need to create new branches in the tree to fulfill the
+			// Condition of one leaf per empty subtree
+			// Note: h is set to the last value from the previous loop
+			let d = binaryKey[h];
+			const currentNodeBinaryKey = binaryExpansion((currentNode as Leaf).key, this.keyLength);
+			let t = currentNodeBinaryKey[h];
+			while (d === t) {
+				// Create branch node with empty value
+				const defaultBranch = new Branch(EMPTY_HASH, EMPTY_HASH);
+				// Append defaultBranch to ancestorNodes
+				ancestorNodes.push(defaultBranch);
+				h += 1;
+				d = binaryKey[h];
+				t = currentNodeBinaryKey[h];
+			}
+			// Create last branch node, parent of node and newLeaf
+			if (+d === 0) {
+				bottomNode = new Branch(newLeaf.hash, currentNode.hash);
+			} else if (+d === 1) {
+				bottomNode = new Branch(currentNode.hash, newLeaf.hash);
+			}
+		}
+		// Finally update all branch nodes in ancestorNodes
+		// Starting from the last
+		while (h > 0) {
+			const d = +binaryKey[h];
+			const p = ancestorNodes[h - 1];
+			if (d === 0 && p instanceof Branch) {
+				// Let siblingNode be the right child node of p
+				const siblingNodeHash = p.rightHash;
+				// Update p.data to bottomNode.hash|siblingNode.hash
+				p.update(bottomNode.hash, NodeSide.LEFT);
+				// Update p.hash to branchHash(p.data)
+				p.update(siblingNodeHash, NodeSide.RIGHT);
+			} else if (d === 1 && p instanceof Branch) {
+				// Let siblingNode be the left child node of p
+				const siblingNodeHash = p.rightHash;
+				// Update p.data to siblingNode.hash|bottomNode.hash
+				p.update(bottomNode.hash, NodeSide.RIGHT);
+				// Update p.hash to branchHash(p.data)
+				p.update(siblingNodeHash, NodeSide.LEFT);
+			}
+			bottomNode = p;
+			h += 1;
```

Comment:
In above loop we have `while (h > 0) {` and here we are doing `h += 1`. It seems loop will never end.

Reply:
ah good catch, its `h -=1` 😅